openwisp_monitoring/monitoring/tests/test_monitoring_notifications.py

```diff
@@ -402,6 +402,132 @@ def test_alerts_disabled(self):
         self.assertEqual(d.monitoring.status, 'problem')
         self.assertEqual(Notification.objects.count(), 0)
 
+    def test_alert_field(self):
+        admin = self._create_admin()
+
+        def _create_alert_field_test_env():
+            m = self._create_general_metric(configuration='test_alert_field')
+            self._create_alert_settings(
+                metric=m, custom_operator='>', custom_threshold=30, custom_tolerance=0
+            )
+            return m
+
+        with self.subTest('Test notification for metric without related field'):
+            m = _create_alert_field_test_env()
+            with self.assertRaises(ValueError) as err:
+                m.write(1)
+            self.assertEqual(
+                str(err.exception),
+                'write() missing keyword argument: "extra_values" required for alert on related field',
+            )
+            m.refresh_from_db()
+            self.assertEqual(m.is_healthy, True)
+            self.assertEqual(m.is_healthy_tolerant, True)
+            self.assertEqual(Notification.objects.count(), 0)
+
+        with self.subTest('Test notification for metric on different related field'):
+            m = _create_alert_field_test_env()
+            with self.assertRaises(ValueError) as err:
+                m.write(10, extra_values={'test_related_3': 40})
+            self.assertEqual(
+                str(err.exception),
+                '"test_related_3" is not defined for alert_field in metric configuration',
+            )
+            m.refresh_from_db()
+            self.assertEqual(m.is_healthy, True)
+            self.assertEqual(m.is_healthy_tolerant, True)
+            self.assertEqual(Notification.objects.count(), 0)
+
+        with self.subTest('Test notification for metric with multiple related fields'):
+            m = _create_alert_field_test_env()
+            m.write(10, extra_values={'test_related_2': 40, 'test_related_3': 20})
+            m.refresh_from_db()
+            self.assertEqual(m.is_healthy, False)
+            self.assertEqual(m.is_healthy_tolerant, False)
+            self.assertEqual(Notification.objects.count(), 1)
+            n = notification_queryset.first()
+            self.assertEqual(n.recipient, admin)
+            self.assertEqual(n.actor, m)
+            self.assertEqual(n.action_object, m.alertsettings)
+            self.assertEqual(n.level, 'warning')
+            n.delete()
```

Comment:
If this sub-test fails, then the notification will not get deleted. And, this would affect the following sub-tests. We should always perform operations related to resetting of the testing environment outside of the sub-tests.

```suggestion
        Notification.objects.all().delete()
```

Same applies for other occurrences.

Reply:
Noted!